File: DP/GridWorldModel.cpp
#include "Support/Error.h"

#include "GridWorldModel.h"
#include "GridWorldPolicy.h"

namespace DP {

std::vector<std::tuple<Reward, Probability, State>> GridWorldModel::Transition(
    State state, Action action) const {
  std::vector<std::tuple<Reward, Probability, State>> returns;
  int x = state / grid_size_, y = state % grid_size_;

  switch (action) {
    case GridWorldAction::up:
      if (x > 0) x -= 1;
      break;
    case GridWorldAction::down:
      if (x < grid_size_ - 1) x += 1;
      break;
    case GridWorldAction::left:
      if (y > 0) y -= 1;
      break;
    case GridWorldAction::right:
      if (y < grid_size_ - 1) y += 1;
      break;
    default:
      eva::FatalError("Unsupported action %d", action);
  }
  returns.push_back(std::make_tuple(-1.0, 1.0, x * grid_size_ + y));
  return returns;
}

bool GridWorldModel::IsTerminalState(State state) const {
  return (state == 0) || (state == grid_size_ * grid_size_ - 1);
}

}  // namespace DP

File: DP/Sources/GridWorld/GridWorldModel.swift
import DynamicProgramming

/// The model for the GridWorld.
///
/// For each action hitting the boundary, it stays.
public class GridWorldModel: Model {
    private let context: Context

    public init(context: Context) {
        self.context = context
    }

    public func isTerminal(for state: State) -> Bool {
        return state.index == 0 || state.index == context.stateCount - 1
    }

    public func feasibleActions(for state: State) -> [Action] {
        assert(!isTerminal(for: state))
        return [
            GridWorldAction.up,
            GridWorldAction.down,
            GridWorldAction.left,
            GridWorldAction.right,
        ]
    }

    public func transition(from state: State, after action: Action) -> [Transition] {
        let stateIndex = state.index
        let gridSize = context.gridSize

        // For GridWorld, we assume the state index is contiguous.
        assert(stateIndex >= 0 && stateIndex < context.stateCount)

        guard let gridWorldAction = action as? GridWorldAction else {
            fatalError("Action must be GridWorldActio.n")
        }

        var x = stateIndex / gridSize
        var y = stateIndex % gridSize

        switch gridWorldAction {
        case .up: if x > 0 { x -= 1 }
        case .down: if x < gridSize - 1 { x += 1 }
        case .left: if y > 0 { y -= 1 }
        case .right: if y < gridSize - 1 { y += 1 }
        }

        return [
            Transition(
                probability: 1.0,
                reward: -1.0,
                newState: GridWorldState(index: x * gridSize + y)
            ),
        ]
    }

    private lazy var modelStates: [State] = {
        var modelStates: [State] = []
        modelStates.reserveCapacity(context.stateCount)
        for i in 0 ..< context.stateCount {
            modelStates.append(GridWorldState(index: i))
        }
        return modelStates
    }()

    public var states: [State] { return modelStates }
}
